### PR TITLE
[RDY] checkhealth: add checkhealth nvim_lsp command

### PR DESCRIPTION
**Pull request by teto**
Only check command for now.

This was much harder than expected as it's dynamically typed, the generation of the config and the start of the client is intertwined, and the logic for merging configs is very good but not the easiest to grasp.

It needs https://github.com/neovim/neovim/pull/11832 to work.

And gives this output:
```
health#nvim_lsp#check
========================================================================
## Checking language server protocol configuration
  - INFO: Checking pyls
  - ERROR: The given command "python" is not executable.
  - INFO: Checking ccls
  - INFO: Checking lua_lsp
  - INFO: Checking ghcide
  - ERROR: The given command "ghcide" is not executable.
  - INFO: Checking texlab
  - ERROR: The given command "texlab" is not executable.
```
It's primitive but it's easy to improve. just check the attributes of new_config.

### CODE REVIEW

File: autoload/health/nvim_lsp.vim [hunk 1]
```diff
@@ -0,0 +1,20 @@
+
+function! health#nvim_lsp#check()
+	call health#report_start('Checking language server protocol configuration')
+lua << EOF
+	local configs = require 'nvim_lsp/configs'
+	local lsp = require 'vim.lsp'
+
+	for name, top_level_config in pairs(configs) do
+		-- the folder needs to exist, => use stdpath ?
+		local new_config = top_level_config.gen_config("/tmp")
```
[the remark this answers is not in the document]
Comment (teto) [reply autoload/health/nvim_lsp.vim]: at first I had used "/fake/path" but the validator checks the folder exists so I was thinking of using `stdpath('cache')` to avoid creating a temporary directory just for checkhealth.  Question is does `stdpath('cache')` always exist ?

**Comment by teto** [comment]
> Can this healthcheck also check for cmds overwritten by the user?

yes the whole make_config merges various defaults with the user config. That's the whole point.

File: autoload/health/nvim_lsp.vim (hunk 1 above)
[the remark this answers is not in the document]
Comment (h-michael) [reply autoload/health/nvim_lsp.vim]: > Question is does stdpath('cache') always exist ?

At least it didn't exist in my environment.
If `stdpath('config')`, init.vim will be placed, so it may not exist in many environments.

@justinmk Is there any good workaround?

**Comment by teto** [comment]
I've tried to address your comments.

File: autoload/health/nvim_lsp.vim (hunk 1 above)
[the remark this answers is not in the document]
Comment (justinmk) [reply autoload/health/nvim_lsp.vim]: > Question is does `stdpath('cache')` always exist ?

Would need to create it here. Is there something wrong with that?

Nvim could add e.g. `ensurecache()`, but that's not different than `mkdir(stdpath('cache'), 'p')`.  Though I suppose Nvim could do this once at startup, then plugins can reasonably assume it exists...

[the remark this answers is not in the document]
Comment (teto) [reply autoload/health/nvim_lsp.vim]: I've been using `.`

**Comment by teto** [comment]
I've written a more elegant version. I just need to test and I push in 10 mn.

File: autoload/health/nvim_lsp.vim [hunk 2]
```diff
@@ -0,0 +1,8 @@
+let s:FILE_DIR = fnamemodify(expand("<sfile>"), ':h') . '/'
+let s:FILE_DIR = substitute(s:FILE_DIR, '\\', '/', 'g')
+
+function! health#nvim_lsp#check()
+	call health#report_start('Checking language server protocol configuration')
+	execute 'luafile '.s:FILE_DIR.'check.lua'
```
Comment (justinmk) [inline autoload/health/nvim_lsp.vim]: if `autoload/health/check.lua` were moved to `lua/nvim_lsp/health.lua`, then this could be:
```suggestion
	require('nvim_lsp.health').nvim_lsp_check_health()
```

And then the workaround in `.luacheckrc` would not be needed.

Is there a reason for the current approach?

File: autoload/health/check.lua [hunk 3]
```diff
@@ -0,0 +1,14 @@
+function nvim_lsp_check_health()
+	local configs = require 'nvim_lsp/configs'
```
Comment (justinmk) [inline autoload/health/check.lua]: we always use 2-space indent.

File: autoload/health/nvim_lsp.vim (hunk 2 above)
Comment (teto) [reply autoload/health/nvim_lsp.vim]: you are right, thats what I've done afterwards but I haven't had the opportunity to test it yet. my system is rebuilding.

File: lua/nvim_lsp/configs.lua [hunk 4]
```diff
@@ -146,6 +147,14 @@ function configs.__newindex(t, config_name, config_def)
...
       end)
+
+      new_config.root_dir = _root_dir
+      return new_config
+    end
+
+    local manager = util.server_per_root_dir_manager(
+    function(_root_dir)
+      local new_config = make_config(_root_dir)
       return new_config
```
Comment (justinmk) [inline lua/nvim_lsp/configs.lua]: strange indent

File: autoload/health/check.lua [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+	for _, top_level_config in pairs(configs) do
+		-- the folder needs to exist
+		local new_config = top_level_config.make_config(".")
+		vim.fn['health#report_info']("Checking "..new_config.name)
+
+		local cmd, _ = vim.lsp._cmd_parts(new_config.cmd)
+		if not (vim.fn.executable(cmd) == 1) then
+			vim.fn['health#report_error'](string.format("The given command %q is not executable.", cmd))
```
Comment (justinmk) [inline autoload/health/check.lua]: 1. brevity
2. the "variable" part of messages should go at the end after a colon, when possible

```suggestion
			vim.fn['health#report_error'](string.format("%s: command is not executable: %q", new_config.name, cmd))
```

File: autoload/health/check.lua [hunk 6]
```diff
@@ -0,0 +1,14 @@
+function nvim_lsp_check_health()
+	local configs = require 'nvim_lsp/configs'
+
+	for _, top_level_config in pairs(configs) do
+		-- the folder needs to exist
+		local new_config = top_level_config.make_config(".")
+		vim.fn['health#report_info']("Checking "..new_config.name)
```
Comment (justinmk) [inline autoload/health/check.lua]: this is noise. instead, show a success message for non-error cases.

File: autoload/health/nvim_lsp.vim (hunk 1 above)
[the remark this answers is not in the document]
Comment (justinmk) [reply autoload/health/nvim_lsp.vim]: isn't it better to use cache or data dirs for their intended purpose?

File: autoload/health/check.lua (hunk 6 above)
Comment (teto) [reply autoload/health/check.lua]: It's difficult to guarantee the success with basic or even more exhaustive checks. 
I just wanted the user to see their config was at least found. So maybe instead: `Found config for X` or `No obvious error for X`.

Comment (justinmk) [reply autoload/health/check.lua]: no need to overthink it. "success" is the opposite of "error".

Comment (teto) [reply autoload/health/check.lua]: I chose `configuration checked`

**Comment by teto** [comment]
Ready. I took the liberty to copy neovim's .editorconfig to prevent future wrong tab/space usage.

File: lua/nvim_lsp/configs.lua [hunk 7]
```diff
@@ -93,7 +93,8 @@ function configs.__newindex(t, config_name, config_def)
       end
       M.manager = nil
     end
-    local manager = util.server_per_root_dir_manager(function(_root_dir)
+
+    local make_config = function(_root_dir)
```
Comment (h-michael) [inline lua/nvim_lsp/configs.lua]:
```suggestion
    local make_config = function(root_dir)
```

File: lua/nvim_lsp/configs.lua [hunk 8]
```diff
@@ -146,8 +147,15 @@ function configs.__newindex(t, config_name, config_def)
...
+
+      new_config.root_dir = _root_dir
       return new_config
-    end)
+    end
+
+    local manager = util.server_per_root_dir_manager(function(_root_dir)
+        local new_config = make_config(_root_dir)
+        return new_config
+      end)
```
Comment (h-michael) [inline lua/nvim_lsp/configs.lua]:
```
    local manager = util.server_per_root_dir_manager(function(root_dir)
      return make_config(root_dir)
    end)
```

File: lua/nvim_lsp/util.lua [hunk 9]
```diff
@@ -208,7 +208,7 @@ end)()
 
 -- Returns a function(root_dir), which, when called with a root_dir it hasn't
 -- seen before, will call make_config(root_dir) and start a new client.
-function M.server_per_root_dir_manager(make_config)
+function M.server_per_root_dir_manager(_make_config)
```
Comment (h-michael) [inline lua/nvim_lsp/util.lua]:
```suggestion
function M.server_per_root_dir_manager(make_config)
```

File: lua/nvim_lsp/util.lua [hunk 10]
```diff
@@ -218,8 +218,7 @@ function M.server_per_root_dir_manager(make_config)
     -- Check if we have a client alredy or start and store it.
     local client_id = clients[root_dir]
     if not client_id then
-      local new_config = make_config(root_dir)
-      new_config.root_dir = root_dir
+      local new_config = _make_config(root_dir)
```
Comment (h-michael) [inline lua/nvim_lsp/util.lua]:
```suggestion
      local new_config = make_config(root_dir)
```

File: lua/nvim_lsp/health.lua [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+
+    local cmd, _ = vim.lsp._cmd_parts(new_config.cmd)
+    if not (vim.fn.executable(cmd) == 1) then
+      vim.fn['health#report_error'](
+        string.format("%s: The given command %q is not executable.", new_config.name, cmd)
+      )
+    else
+      vim.fn['health#report_info'](
+        string.format("%s: configuration checked.", new_config.name)
+      )
```
Comment (h-michael) [inline lua/nvim_lsp/health.lua]: I think this is better, what do you think? cc: @justinmk 
```
      vim.fn['health#report_ok'](
        string.format("%s: The given command %q is executable.", new_config.name, cmd)
      )
```

File: lua/nvim_lsp/configs.lua [hunk 12]
```diff
@@ -146,8 +147,15 @@ function configs.__newindex(t, config_name, config_def)
               ))
         end
       end)
+
+      new_config.root_dir = _root_dir
```
Comment (h-michael) [inline lua/nvim_lsp/configs.lua]: Why did you move this process here? I don't think I need to move.

File: lua/nvim_lsp/health.lua [hunk 13]
```diff
@@ -0,0 +1,22 @@
+local M = {}
+function M.check_health()
+  local configs = require 'nvim_lsp/configs'
+
+  for _, top_level_config in pairs(configs) do
+    -- the folder needs to exist
+    local new_config = top_level_config.make_config(".")
```
Comment (h-michael) [inline lua/nvim_lsp/health.lua]: If you do not [move this process](https://github.com/neovim/nvim-lsp/pull/120/files#diff-bb1a5a30518e3cb9d6bb43af411ddbf5R151), it should work without any arguments.

File: lua/nvim_lsp/util.lua (hunk 9 above)
Comment (teto) [reply lua/nvim_lsp/util.lua]: This avoids collisions with the function `make_config`. I prefer it the current way,

File: lua/nvim_lsp/configs.lua (hunk 12 above)
Comment (teto) [reply lua/nvim_lsp/configs.lua]: because I use to call `validate_client_config` and it would complain about a missing root_dir. make_config should generate a full config and not a half decent one. Because I don't use `validate_client_config` anymore, I can now pass a fake path (lsp will validate it only upon starting lsp) so it solves the issue of having a valid path. Thanks for making me notice.

File: lua/nvim_lsp/health.lua (hunk 11 above)
Comment (teto) [reply lua/nvim_lsp/health.lua]: If we add more checks later (I am sure we will), it's best to have a not too detailed message for success.

**Comment by teto** [comment]
thanks for the reviews
